service/matching/matchingEngine.go

```diff
@@ -321,6 +336,59 @@ pollLoop:
 	}
 }
 
+// QueryWorkflow creates a DecisionTask with query data, send it through sync match channel, wait for that DecisionTask
+// to be processed by worker, and then return the query result.
+func (e *matchingEngineImpl) QueryWorkflow(ctx context.Context, queryRequest *m.QueryWorkflowRequest) (*workflow.QueryWorkflowResponse, error) {
+	domainID := queryRequest.GetDomainUUID()
+	taskListName := *queryRequest.TaskList.Name
+	taskList := newTaskListID(domainID, taskListName, persistence.TaskListTypeDecision)
+	tlMgr, err := e.getTaskListManager(taskList)
+	if err != nil {
+		return nil, err
+	}
+	queryTask := &queryTaskInfo{
+		queryRequest: queryRequest,
+		taskID:       uuid.New(),
+	}
+	err = tlMgr.SyncMatchQueryTask(ctx, queryTask)
+	if err != nil {
+		return nil, err
+	}
+
+	queryResultCh := make(chan *workflow.RespondQueryTaskCompletedRequest)
+	e.queryMapLock.Lock()
+	e.queryTaskMap[queryTask.taskID] = queryResultCh
+	e.queryMapLock.Unlock()
+	defer func() {
+		e.queryMapLock.Lock()
+		delete(e.queryTaskMap, queryTask.taskID)
+		e.queryMapLock.Unlock()
+	}()
+
+	select {
+	case result := <-queryResultCh:
+		if *result.CompletedType == workflow.QueryTaskCompletedTypeFailed {
+			return nil, &workflow.QueryFailedError{Message: result.GetErrorMessage()}
+		}
+		return &workflow.QueryWorkflowResponse{QueryResult: result.QueryResult}, nil
+	case <-ctx.Done():
```

Comment:
I think you need to check the error:
if ctx.Err() == context.DeadlineExceeded 
to make sure it is timeout.

Reply:
but what if it is not, what could we do differently here?